cmd/aws-vpc-cni/main.go

```diff
@@ -196,39 +204,84 @@ func waitForInit() error {
+func getPrimaryIP(ipv4 bool) (string, error) {
 	var hostIP string
 	var err error
-	for {
-		hostIP, err = imds.GetMetaData("local-ipv4")
-		if err != nil {
-			log.WithError(err).Fatalf("aws-vpc-cni failed")
-			return "", err
-		}
-		if hostIP != "" {
-			return hostIP, nil
+	var key string
+
+	if ipv4 {
+		key = "local-ipv4"
+	} else {
+		key = "ipv6"
+	}
+	hostIP, err = cniutils.GetNodeMetadata(key)
+	if err != nil {
+		if ipv4 {
+			log.WithError(err).Fatalf("failed to rtrieve local-ipv4 address in imds metadata")
+		} else {
+			log.WithError(err).Debugf("failed to retrieve ipv6 address in imds metadata")
 		}
+		return "", err
 	}
+	return hostIP, nil
 }
 
 func isValidJSON(inFile string) error {
 	var result map[string]interface{}
 	return json.Unmarshal([]byte(inFile), &result)
 }
-
-func generateJSON(jsonFile string, outFile string, nodeIP string) error {
+func generateJSON(jsonFile string, outFile string, getPrimaryIP func(ipv4 bool) (string, error)) error {
 	byteValue, err := os.ReadFile(jsonFile)
 	if err != nil {
 		return err
 	}
 
+	// enabledIPv6 is to determine this EKS cluster is IPv4 or IPv6 cluster
+	// if this EKS cluster is IPv6 cluster, egress-cni-plugin will enable IPv4 egress by default
+	// if this EKS cluster is IPv4 cluster, egress-cni-plugin will enable IPv6 egress by default
+	enabledIPv6 := utils.GetBoolAsStringEnvVar(envEnIPv6, defaultEnableIPv6)
+	var egressIPAMSubnet string
+	var egressIPAMDst string
+	var egressIPAMDataDir string
+	var egressEnabled bool
+	var egressPluginLogFile string
+	var nodeIP = ""
+	if enabledIPv6 {
+		// EKS IPv6 cluster
+		egressIPAMSubnet = egressPluginIpamSubnetV4
+		egressIPAMDst = egressPluginIpamDstV4
+		egressIPAMDataDir = egressPluginIpamDataDirV4
+		egressEnabled = true
+		egressPluginLogFile = utils.GetEnv(envEgressV4PluginLogFile, defaultEgressV4PluginLogFile)
+		nodeIP, err = getPrimaryIP(true)
+		// Node should have a IPv4 address even in IPv6 cluster
+		if err != nil {
+			log.Errorf("Failed to get Node IP, error: %v", err)
+			return err
+		}
+	} else {
+		// EKS IPv4 cluster
+		egressIPAMSubnet = egressPluginIpamSubnetV6
+		egressIPAMDst = egressPluginIpamDstV6
+		egressIPAMDataDir = egressPluginIpamDataDirV6
+		egressEnabled = utils.GetBoolAsStringEnvVar(envEnIPv6Egress, defaultEnableIPv6Egress)
+		egressPluginLogFile = utils.GetEnv(envEgressV6PluginLogFile, defaultEgressV6PluginLogFile)
+		if egressEnabled {
+			nodeIP, err = getPrimaryIP(false)
+			// Node may not have a IPv6 address if there is no need to support IPv6 egress in eks IPv4 cluster
+			if err != nil {
+				log.Errorf("To support IPv6 egress, node primary ENI must have a global IPv6 address, error: %v", err)
+				// Global unicast IPv6 nodeIP is mandatory to support IPv6 egress traffic in EKS IPv4 cluster
+				// no return err here so that aws-node daemonset pod could start success
+				// chained egress-cni-plugin will emit error message about empty nodeIP if IPv6 egress is enabled
```

Comment:
I think we can return error here, as the combination of setting `ENABLE_V6_EGRESS` and not configuring an IPv6 node IP is an error

Reply:
changed.